// @clayui/core/lib/collection/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Collection", {
  enumerable: true,
  get: function get() {
    return _Collection.Collection;
  }
});
Object.defineProperty(exports, "excludeProps", {
  enumerable: true,
  get: function get() {
    return _utils.excludeProps;
  }
});
Object.defineProperty(exports, "getKey", {
  enumerable: true,
  get: function get() {
    return _utils.getKey;
  }
});
Object.defineProperty(exports, "useCollection", {
  enumerable: true,
  get: function get() {
    return _useCollection.useCollection;
  }
});
Object.defineProperty(exports, "useCollectionKeys", {
  enumerable: true,
  get: function get() {
    return _useCollection.useCollectionKeys;
  }
});
Object.defineProperty(exports, "useVirtual", {
  enumerable: true,
  get: function get() {
    return _useVirtual.useVirtual;
  }
});

var _utils = require("./utils");

var _Collection = require("./Collection");

var _useCollection = require("./useCollection");

var _useVirtual = require("./useVirtual");
